src/zepben/evolve/services/network/tracing/networktrace/network_trace_step_path_provider.py

```diff
@@ -0,0 +1,273 @@
+            # We also make sure we filter out the cut or the clamp we are starting at, so we don't compare it in our checks
+            filter_func = lambda it: it != from_terminal.conducting_equipment and self.state_operators.is_in_service(it)
+            cuts: List[Cut] = list(filter(filter_func, cuts))
+            clamps: List[Clamp] = list(filter(filter_func, clamps))
+
+            cuts_at_same_position = list(filter(lambda it: it.length_from_T1_or_0 == length_from_T1, cuts))
+            stop_at_cuts_at_same_position = bool(can_stop_at_cut_at_same_position and cuts_at_same_position)
+
+            def next_cut_length_from_terminal_1_func():
+                if stop_at_cuts_at_same_position:
+                    return length_from_T1
+                elif towards_segment_T2:
+                    return min((it.length_from_T1_or_0 for it in cuts if it.length_from_T1_or_0 > length_from_T1), default=None)
+                else:
+                    return max((it.length_from_T1_or_0 for it in cuts if it.length_from_T1_or_0 < length_from_T1), default=None)
+
+            next_cut_length_from_terminal_1 = next_cut_length_from_terminal_1_func()
+
+            next_cuts = [it for it in cuts if it.length_from_T1_or_0 == next_cut_length_from_terminal_1] if next_cut_length_from_terminal_1 is not None else []
+
+            def next_term_length_from_term_1_func():
+                if next_cut_length_from_terminal_1 is not None:
+                    return next_cut_length_from_terminal_1
+                elif towards_segment_T2:
+                    return acls_length_or_max(acls)
+                else:
+                    return 0.0
+
+            next_terminal_length_from_terminal_1 = next_term_length_from_term_1_func()
+
+            def clamps_before_next_terminal_filter() -> Callable[[Clamp], bool]:
+                if isinstance(from_terminal.conducting_equipment, AcLineSegment) and towards_segment_T2:
+                    return lambda it: length_from_T1 <= it.length_from_T1_or_0 <= next_terminal_length_from_terminal_1
+                elif towards_segment_T2:
+                    return lambda it: it.length_from_T1_or_0 > length_from_T1 and it.length_from_T1_or_0 <= next_terminal_length_from_terminal_1
+                elif (next_terminal_length_from_terminal_1 == 0.0) and len(next_cuts) == 0:
+                    return lambda it: next_terminal_length_from_terminal_1 <= it.length_from_T1_or_0 <= length_from_T1
+                else:
+                    return lambda it: it.length_from_T1_or_0 <= length_from_T1  and it.length_from_T1_or_0 > next_terminal_length_from_terminal_1
+            _filter = clamps_before_next_terminal_filter()
+
+            clamps_before_next_terminal = filter(_filter, clamps)
+
+            next_stop_terminals = [] if stop_at_cuts_at_same_position else (
+                    it.get_terminal(1 if towards_segment_T2 else 2) for it in next_cuts
```

Comment:
question: I notice you do this in a lot of places. Is it just a personal preference to collect to a tuple rather than a list, or is there a performance (or some other) benefit for doing so?